test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/WriteJsonPayloadTests/SerializationInfoEdgeTests.cs

```diff
@@ -0,0 +1,488 @@
+
+                await odataWriter.WriteStartAsync(feed);
+                await odataWriter.WriteStartAsync(entry);
+                await odataWriter.WriteEndAsync();
+                await odataWriter.WriteEndAsync();
+                Stream stream = await responseMessageWithoutModel.GetStreamAsync();
+                string result = await TestsHelper.ReadStreamContentAsync(stream);
+                if (!mimeType.Contains(MimeTypes.ODataParameterNoMetadata))
+                {
+                    Assert.Contains(NameSpacePrefix + "Employee", result);
+                    Assert.Contains("$metadata#People", result);
+                }
+                else
+                {
+                    Assert.DoesNotContain(NameSpacePrefix + "Employee", result);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Write response payload with serialization info containing wrong values
+        /// </summary>
+        [Theory]
+        [InlineData(MimeTypeODataParameterFullMetadata)]
+        [InlineData(MimeTypeODataParameterMinimalMetadata)]
+        [InlineData(MimeTypeODataParameterNoMetadata)]
+        public async Task WriteEntryWithWrongSerializationInfo(string mimeType)
+        {
+            var settings = new ODataMessageWriterSettings();
+            settings.ODataUri = new ODataUri() { ServiceRoot = _baseUri };
+
+            // wrong EntitySetName for entry
+            var responseMessageWithoutModel = new TestStreamResponseMessage(new MemoryStream());
+            responseMessageWithoutModel.SetHeader("Content-Type", mimeType);
+            using (var messageWriter = new ODataMessageWriter(responseMessageWithoutModel, settings))
+            {
+                var odataWriter = await messageWriter.CreateODataResourceWriterAsync();
+                var entry = this.CreatePersonEntryWithoutSerializationInfo();
+
+                entry.SetSerializationInfo(new ODataResourceSerializationInfo() { NavigationSourceName = "Parsen", NavigationSourceEntityTypeName = NameSpacePrefix + "Person" });
+                await odataWriter.WriteStartAsync(entry);
+                await odataWriter.WriteEndAsync();
+                var result = await TestsHelper.ReadStreamContentAsync(responseMessageWithoutModel.GetStream());
+                Assert.Contains("\"PersonId\":-5", result);
+                if (!mimeType.Contains(MimeTypes.ODataParameterNoMetadata))
+                {
+                    // no metadata does not write odata.metadata
+                    Assert.Contains("$metadata#Parsen/$entity", result);
+                    Assert.Contains("People(-5)\",\"PersonId\":-5", result);
+                }
+                else
+                {
+                    Assert.DoesNotContain("People(-5)\",\"PersonId\":-5", result);
+                }

```

Comment:
Consider avoiding conditions that make it harder to tell what is being tested when. We could split the test or parameterize the test to avoid conditions.

Reply:
I've split these tests